Implement Content Security Policy, including moving all CSS and JS to external files

candrews opened the pull request:
A strict content security policy is increasingly required and expected. This PR allows JavaMelody to function when a strict CSP is set, and in also sets a strict CSP for JavaMelody itself by default.

I've tested this pretty well on my own system, clicking around and checking various things to ensure everything looks and works correctly. There are no functional or visual changes; if I've done this right, you shouldn't be able to tell that I've done anything at all. :)

* Add a configuration option for CSP header (enabled by default) 
* Eliminate all inline script and styles 

Completes the work started in https://github.com/javamelody/javamelody/issues/782

candrews commented:
Build error is unrelated. That error is addressed by https://github.com/javamelody/javamelody/pull/965

candrews commented:
@evernat have you had a chance to look at this PR? I'm eager to address any feedback in order to get this work merged - thank you again!

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/Parameter.java:
```diff
@@ -202,7 +202,7 @@
 
 	/**
-	 * Active la protection contre CSRF (false par défaut).
+	 * Parameter to enable or disable the Content Security Policy header (true by default).
```
This is not the right place for this javadoc.

candrews replied:
Fixed!

evernat commented:
I looked at the PR, thank you for that. There is much to read and to test.

candrews commented:
> There is much to read and to test.

Yeah, it's a big one...

Is there anything I can do to help? As I mentioned, I'm quite eager to get this merged :)

evernat commented:
The PR is not bad, but there are things broken by the PR. It needs tests and I will try to list some issues.

candrews commented:
> The PR is not bad

:tada: 


> but there are things broken by the PR.

:( I'm eager to eager what so I can address those issues.



> It needs tests and I will try to list some issues.

Do you mean some kind of automated tests are required? I don't see any testing framework in JavaMelody.

evernat commented:
When having Maven versions in the webapp and when choosing a period by version at the top of the main report, it does nothing. It should refresh the page with the period of the version.
After opening the "Database" report in "System informations", choosing a database report in the list does nothing. It should refresh the page to display the choosen report.
Both are because the browser blocks inline scripts still written with "onchange".

evernat commented:
> Do you mean some kind of automated tests are required? I don't see any testing framework in JavaMelody.

There is junit. But given it's JS and CSP, I was speaking of manual tests.

evernat commented:
When clicking on "Execute the garbage collector", the confim message displays "<br>" instead of a line break.

evernat commented:
In the Threads details, clicking on the "Send a thread interrupt signal" or "Kill the tread" actions display a confirm message "null" instead of "Do you confirm ..." and "... Do you really want ..."

candrews commented:
> When having Maven versions in the webapp and when choosing a period by version at the top of the main report, it does nothing. It should refresh the page with the period of the version.
> After opening the "Database" report in "System informations", choosing a database report in the list does nothing. It should refresh the page to display the choosen report.
> Both are because the browser blocks inline scripts still written with "onchange".

I missed `onchange` - thank you for pointing those out. I've now addressed them.

> When clicking on "Execute the garbage collector", the confim message displays `<br>` instead of a line break.

Fixed

> In the Threads details, clicking on the "Send a thread interrupt signal" or "Kill the tread" actions display a confirm message "null" instead of "Do you confirm ..." and "... Do you really want ..."

Fixed

I've now addressed everything you reported; if you have any more findings or other feedback, please let know!

candrews commented:
Happy 2021! :partying_face: 

I have my :crossed_fingers: that the merging of this PR will be a great new start to the new year!

candrews commented:
@evernat could you please take another look at this PR?

candrews commented:
Hello @evernat - please let me know if there's anything I can do to help this PR along

evernat commented:
I have made some tests and there are some issues:

1. In the database reports, the list to choose the database report does not work, That's because there is `&amp;` in the URL instead of `&`. See http://javamelody.org/demo/monitoring?part=database
2. The `Other charts` link on the right does not work well: it displays the other charts but the items do not scroll below the charts like in http://javamelody.org/demo/monitoring.
3. The slider to zoom the graph does not work like it does in http://javamelody.org/demo/monitoring?part=graph&graph=usedMemory. The image should be zoomed with a new URL like http://javamelody.org/demo/monitoring?graph=usedMemory&width=1480&height=662
4. When clicking on Details in System information, just above Threads, the Threads part slips on the right, but it shoud not, see http://javamelody.org/demo/monitoring

candrews commented:
> I have made some tests and there are some issues:
> 
>     1. In the database reports, the list to choose the database report does not work, That's because there is `&amp;` in the URL instead of `&`. See http://javamelody.org/demo/monitoring?part=database

I've pushed a fix for this issue.
> 
>     2. The `Other charts` link on the right does not work well: it displays the other charts but the items do not scroll below the charts like in http://javamelody.org/demo/monitoring.


I've pushed a fix for this issue.
> 
>     3. The slider to zoom the graph does not work like it does in http://javamelody.org/demo/monitoring?part=graph&graph=usedMemory. The image should be zoomed with a new URL like http://javamelody.org/demo/monitoring?graph=usedMemory&width=1480&height=662

I can't reproduce this one. I moved the slider and I can see that the image src keep changing, and does do in the same way as the demo site does.
> 
>     4. When clicking on Details in System information, just above Threads, the Threads part slips on the right, but it shoud not, see http://javamelody.org/demo/monitoring

I've pushed a fix for this issue.

Thanks again for pointing out these issue - hopefully we're close to being able to merge this now :crossed_fingers:

candrews commented:
@evernat friendly bump - I'd love to make progress and get this change merged, thanks again!

candrews commented:
@evernat is there anything I can do to help this along?

evernat commented:
Sorry for the delay.
I have made some test and I have some issues:
1. When I see the parameter `content-security-policy` I would think that the value should be `default-src 'self'; object-src 'none';`. So it may be better to rename it to `content-security-policy-enabled`. (true by default)
2. When I click on some actions such as Run GC. There is no more message giving the result of the action. It suppose that it is related to `htmlCoreReport.toHtml(message, anchorNameForRedirect);` replaced by `htmlCoreReport.toHtml();` in the PR.
3. When the monitoring page is loaded, graphs of http and sql requests for example are loaded with URL like .../monitoring?graph=http211e03eeda2e3305b34429bf203bbe8e280d4664&width=100&height=50. Those graphs of http and sql requests should be loaded only when the mouse hovers over each request in the details in http and sql statistics. It is related to `<img src='?graph=` and the removing of the onmouseover script in the PR. That's important because on some applications there may up to 10000 http requests and 10000 sql requests in the statistics.
4. There are at least one issue with content-security-policy: in HtmlMBeansReport, there is still an inline `<style type='text/css'>`

candrews commented:
> * When I see the parameter `content-security-policy` I would think that the value should be `default-src 'self'; object-src 'none';`. So it may be better to rename it to `content-security-policy-enabled`. (true by default)

Good point. I've updated the PR accordingly.

> 
> * When I click on some actions such as Run GC. There is no more message giving the result of the action. It suppose that it is related to `htmlCoreReport.toHtml(message, anchorNameForRedirect);` replaced by `htmlCoreReport.toHtml();` in the PR.

I'll work on that...

> 
> * When the monitoring page is loaded, graphs of http and sql requests for example are loaded with URL like .../monitoring?graph=http211e03eeda2e3305b34429bf203bbe8e280d4664&width=100&height=50. Those graphs of http and sql requests should be loaded only when the mouse hovers over each request in the details in http and sql statistics. It is related to `<img src='?graph=` and the removing of the onmouseover script in the PR. That's important because on some applications there may up to 10000 http requests and 10000 sql requests in the statistics.

I'll work on that...

> 
> * There are at least one issue with content-security-policy: in HtmlMBeansReport, there is still an inline `<style type='text/css'>`

Thanks for finding that - I've now fixed it.

candrews commented:
@evernat I've now addressed all of the newly reported issues and I'm not aware of anything else.

So please let me know what else I can do. Or, perhaps this can now be merged? :angel: 

Thanks again!

candrews commented:
@evernat can you please take another look?

candrews commented:
I'm sorry for the pinging, I just really want to get this PR done and off my todo list. I have projects that really want this functionality and my colleagues are waiting for it.

Please let me know what I can do to get this across the finish line and into a release of javamelody.

Thank you once again for your work and continued maintenance of this very helpful, very useful project.

evernat commented:
I have tested and core seems ok at the moment.
When used in the collect server, I have 2 issues:
- CollectorController.showAlertAndRedirect does not work anymore (the page is only a span without doctype, html, head, script or body tag
- CollectorController.writeDataUnavailableForApplication : the confirm dialog is not displayed anymore

I was looking how to fix that.

evernat commented:
For the reason above, I suggest the following methods in CollectorController:

```java
	public void writeMessage(HttpServletRequest req, HttpServletResponse resp, String message)
			throws IOException {
		if (HttpParameter.PART.getParameterFrom(req) == null) {
			showAlertAndRedirectTo(resp, message, "?");
		} else {
			final String partToRedirectTo;
			if (HttpParameter.CACHE_ID.getParameterFrom(req) == null) {
				partToRedirectTo = HttpParameter.PART.getParameterFrom(req);
			} else {
				partToRedirectTo = HttpParameter.PART.getParameterFrom(req) + '&'
						+ HttpParameter.CACHE_ID + '='
						+ HttpParameter.CACHE_ID.getParameterFrom(req);
			}
			showAlertAndRedirectTo(resp, message, "?part=" + partToRedirectTo);
		}
	}

	public static void writeDataUnavailableForApplication(String application,
			HttpServletResponse resp) throws IOException {
		final PrintWriter writer = createWriterFromOutputStream(resp);
		writer.write("<html lang='" + I18N.getCurrentLocale().getLanguage() + "'>");
		writer.write("<head><title>Monitoring</title>");
		writer.write("<script type='text/javascript' src='?resource=prototype.js'></script>");
		writer.write("<script type='text/javascript' src='?resource=monitoring.js'></script>");
		writer.write("</head><body>");
                ...
        }

	public static void showAlertAndRedirectTo(HttpServletResponse resp, String message,
			String redirectTo) throws IOException {
		final PrintWriter writer = createWriterFromOutputStream(resp);
		writer.write("<html lang='" + I18N.getCurrentLocale().getLanguage() + "'>");
		writer.write("<head><title>Monitoring</title>");
		writer.write("<script type='text/javascript' src='?resource=prototype.js'></script>");
		writer.write("<script type='text/javascript' src='?resource=monitoring.js'></script>");
		writer.write("</head><body>");
		writer.write("<span class='alertAndRedirect' data-alert='"
				+ I18N.htmlEncode(message, false, false) + "' data-href='" + redirectTo
				+ "'></span>");
		writer.write("</body></html>");
		writer.close();
	}
```

And when displaying an alert dialog or I suppose a confirm dialog, quotes in the message are an issue: the message is truncated before the end. Previously, quotes and others were escaped by what is now urlEncode.

candrews commented:
> When used in the collect server, I have 2 issues:
> 
>     * CollectorController.showAlertAndRedirect does not work anymore (the page is only a span without doctype, html, head, script or body tag)
> 
>     * CollectorController.writeDataUnavailableForApplication : the confirm dialog is not displayed anymore
> 

I fixed those issues; there were a few problems.

* Collector Server wasn't serving resources, so I change it do that.
* I escaped single and double quotes when escaping html
* CollectorController.showAlertAndRedirectTo now returns a full html page
* A few other bug fixes

The collector server now seems to work correctly and display error messages.

candrews commented:
@evernat how are things now? Is there anything else you see as problematic that I can address?

evernat commented:
Hi,
There are still issues with encoding. In particular, StringIndexOutOfBoundsException when a confirm message contains a quote, now encoded as `&#39;`, because keys between `#` are intended to be translations keys in write*. For example, in the confirm message for GC in French.
Ignoring `&#` in the translation code is not a solution, because there would be `&#39;` in confirm messages anyway.

So my plan is:
- I cut a release now, because it is overdue by too many months. I can still make another release when the content-security-policy work is ready.
- I have merged the PR in the content-security-policy branch for reference.
Then probably:
- I merge some bits such as moving some inline styles to css and moving some inline scripts to js files, to make the PR to come more readable. Of course, I say in the commits your github id and that it is based on this PR. And I rebase the branch with that.
- I look again at which encoding methods are used where and what is broken by what, in order to fix the branch.
- I add something like writeHtmlBegin before writeMessageIfNotNull in CollectorController.writeMessage.

The work in this PR is already huge and probably the bigger I have seen in a PR for javamelody. It's not bad if it is not ready yet given the task that it is.

candrews commented:
The `#` issue can be easily fixed by using a different character representation. I've submitted a PR at https://github.com/javamelody/javamelody/pull/1003 that does just that.

I'm sorry I keep having issues... without comprehensive tests and since I'm not fully familiar with all of the functionality across the entire interface, it's really hard for me to find all the problems. I'm trying though!

Do you have any other issues? I'll happily fix them.

If you'd like the commits rebased for clarity, I can do that. I was (and still am) unclear on if you intend to squash all the commits.

I really want to get this into JavaMelody. I've been trying for almost 6 months; it's important to myself and my colleagues. Can we do anything to accelerate the process?

evernat commented:
Given the number of commits in the branch, yes the commits would be squashed to be merged.
But I plan first to commit some css and js changes from the branch and then to rebase the branch.
The branch will be merged after a final review.

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlJCacheInformationsReport.java:
```diff
@@ -197,16 +198,16 @@ private void writeJCacheKeys(JCacheInformations jcacheInformations) throws IOExc
 				writeDirectly("</td>");
 				if (systemActionsEnabled) {
-					writeDirectly("<td class='noPrint' style='text-align: center;'>");
+					writeDirectly("<td class='noPrint containingIcon'");
```
`>` was missing here

candrews replied:
Nice catch, thank you!

srsouto commented:
Is there a status on the merge of the rest of this branch? I'm seeing that the css and js externalize work was merged, but has the CSP implementation been pushed through?

tonylulciuc commented:
@evernat Do you know if this has been pushed through?

candrews commented:
This hasn't actually been merged to master - it's been "merged" to a branch: https://github.com/javamelody/javamelody/tree/content-security-policy

I don't know if or when @evernat intends to merge this work to master and include it in a release, nor do I know what's blocking it. I've been trying to get this done for 7 months but @evernat is very slow to respond and I'm honestly not sure how to progress this. I'm starting to think CSP will simply never happen in JavaMelody, despite my dedication to it. :-(

evernat commented:
Generally speaking, you could note that my time to work on that is not infinite. And that it could not be merged if there are known issues or if after many reviews / fixes, review for merge is needed to be sure to avoid blocking regressions for a lot of users. See PR #1012 for example, where fixes of about one line of code out of five has a review, just to scratch the surface. In this PR #966 thanks to @candrews, work and changes were risky and big as the needed reviews. You may look at other major opensource projects, where most issues are not answered for years and closed without explanation, to compare with javamelody.
And I think I will have more time to work on that, starting 10 days from now.

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlCacheInformationsReport.java:
```diff
@@ -54,9 +54,9 @@ void toHtml() throws IOException {
-					+ "' onclick=\"javascript:return confirm('"
-					+ getStringForJavascript("confirm_purge_caches") + "');\">");
+			writeln("<a class='confirm' href='?action=clear_caches" + getCsrfTokenUrlPart()
+					+ "' data-confirm=\"" + I18N.htmlEncode("confirm_purge_caches", false, false)
```
The message is not translated anymore here.

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlCacheInformationsReport.java:
```diff
@@ -217,16 +218,16 @@ private void writeCacheKeys(CacheInformations cacheInformations) throws IOExcept
 					writeDirectly(urlEncode(myKey));
 					writeDirectly(csrfTokenUrlPart);
-					writeDirectly("' onclick=\"javascript:return confirm('");
+					writeDirectly("' data-confirm\"");
```
`=` is missing here

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlJCacheInformationsReport.java:
```diff
@@ -197,16 +198,16 @@ private void writeJCacheKeys(JCacheInformations jcacheInformations) throws IOExc
+					writeDirectly("' data-confirm=\"");
 					writeDirectly(confirmClearCache);
-					writeDirectly("');\">");
+					writeDirectly("\");\">");
```
`\");` should not be here

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlJavaInformationsReport.java:
```diff
@@ -363,19 +363,9 @@ private void writeGraph(String graph, String value) throws IOException {
+		write("'>");
+		write("<em><img src='?graph=" + graph + "&width=100&height=50' id='");
+		write("id" + graph);
 		write("' alt='graph'/></em>");
```
after this change, these images are no longer lazy loaded when onmouseover

evernat commented on javamelody-core/src/main/resources/net/bull/javamelody/resource/monitoring.js:
```diff
@@ -0,0 +1,237 @@
+	  });
+
+	  $$('.deploymentPeriod').invoke("observe", "click", function(event){
+		showHide('customPeriod');
```
if this was used, it would have displayed the custom period form instead of the deployment period form

evernat commented on javamelody-core/src/main/resources/net/bull/javamelody/resource/monitoring.js:
```diff
@@ -0,0 +1,237 @@
+		var test = document.createElement('input'); test.type = 'date';
+		if(test.type === 'text') {
+		  document.customPeriodForm.pattern.value = '';
+		  document.getElementById('customPeriodPattern').style.display='inline';
```
it would have been better to not remove comments on this code

evernat commented on javamelody-core/src/main/resources/net/bull/javamelody/resource/monitoring.js:
```diff
@@ -0,0 +1,237 @@
+		      return false;
+		   }
+		   if (periodForm.endDate.value.length == 0) {
+		      alert('Dates are mandatory');
```
This is no longer translated. Moreover the translations for "Dates are mandator" are left in properties files.

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlReport.java:
```diff
@@ -212,62 +210,24 @@ private void writeHtmlHeader(boolean includeSlider, boolean includeCssInline)
 		if (includeSlider) {
 			writeln("<script type='text/javascript' src='?resource=slider.js'></script>");
 		}
+		writeln("<script type='text/javascript' src='?resource=monitoring.js'></script>");
```
when includeCssInline above, includes of prototype.js and monitoring.js are missing

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/CollectorController.java:
```diff
@@ -537,11 +545,11 @@ public static void writeDataUnavailableForApplication(String application,
-		writer.write("');location.href='");
-		writer.write(redirectTo);
-		writer.write("';</script>");
+		writeHtmlBegin(writer);
```
The same `writeHtmlBegin(writer);` is missing in the writeMessage method.

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlJobInformationsReport.java:
```diff
@@ -76,15 +76,17 @@ void toHtml() throws IOException {
 					+ endOnClickConfirm);
 			writeln("<img src='?resource=control_pause_blue.png' width='18' height='18' alt=\"#Pause_all_jobs#\" /> #Pause_all_jobs#</a>");
 			writeln("&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;");
 			writeln("<a href='?action=resume_job&amp;jobId=all" + getCsrfTokenUrlPart()
```
`class='confirm'` is missing here

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlJobInformationsReport.java:
```diff
@@ -206,16 +208,18 @@ private void writeJobTimes(JobInformations jobInformations, CounterRequest count
-		final String endOnClickConfirm = "');\">";
+		final String onClickConfirm = "' data-confirm=\"";
+		final String endOnClickConfirm = "\">";
 		writeln("<a href='?action=pause_job&amp;jobId=" + jobInformations.getGlobalJobId()
```
`class='confirm'` is also missing here

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlJobInformationsReport.java:
```diff
@@ -206,16 +208,18 @@ private void writeJobTimes(JobInformations jobInformations, CounterRequest count
+				+ endOnClickConfirm);
 		writeln("<img src='?resource=control_pause_blue.png' width='18' height='18' alt=\"#Pause_job#\" title=\"#Pause_job#\" /></a>");
 		write("</td> <td align='center' class='noPrint'>");
 		writeln("<a href='?action=resume_job&amp;jobId=" + jobInformations.getGlobalJobId()
```
`class='confirm'` is also missing here

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlHeapHistogramReport.java:
```diff
@@ -173,21 +173,21 @@ private void writeLinks() throws IOException {
-					+ getStringForJavascript("ramasse_miette_desactive") + "');return false;\">");
+			writeln("<a class='alert' href='?part=heaphisto&amp;action=gc" + getCsrfTokenUrlPart()
+					+ "' data-alert=\""
+					+ I18N.htmlEncode(getString("ramasse_miette_desactive"), false, false) + "\">");
```
alert is not displayed instead of calling the action

evernat replied:
by lack of case in monitoring.js

evernat commented on javamelody-core/src/main/java/net/bull/javamelody/internal/web/html/HtmlCoreReport.java:
```diff
@@ -725,20 +691,19 @@ private void writeSystemActionsLinks() throws IOException { // NOPMD
-			write("<a href='?action=gc" + getCsrfTokenUrlPart() + "' onclick=\"javascript:alert('"
-					+ getStringForJavascript("ramasse_miette_desactive") + "');return false;\">");
+			write("<a class='alert' href='?action=gc" + getCsrfTokenUrlPart() + "' data-alert=\""
+					+ I18N.htmlEncode(getString("ramasse_miette_desactive"), false, false) + "\">");
```
alert is not displayed here neither instead of calling the action
